ModelViewer/ModelViewer/ViewerGraphicsWindow.cpp

```diff
@@ -57,6 +57,136 @@ bool ViewerGraphicsWindow::loadModel(QString filepath) {
     return m_currentModel.m_isValid;
 }
 
+bool ViewerGraphicsWindow::loadVertexShader(QString vertfilepath)
+{
+    if (!initialized)
+    {
+        return false;
+    }
+    if (vertfilepath.isEmpty()) {
+        vertfilepath = QFileDialog::getOpenFileName(nullptr, "Load Vertex Shader", "../Data/Shaders/", "");
+        if (vertfilepath.isEmpty()) {
+            return false;
+        }
+    }
+
+    //QOpenGLShaderProgram *newProgram = new QOpenGLShaderProgram(this);
+    m_program->removeAllShaders();
+
+    m_program->addShaderFromSourceFile(QOpenGLShader::Vertex, vertfilepath);
+    m_program->addShaderFromSourceFile(QOpenGLShader::Fragment, currentFragFile);
+    if (!m_program->link())
+    {
+        QString error = m_program->log();
+        qDebug() << error << endl;
+        currentVertFile = vertfilepath;
+        //m_program->addShaderFromSourceFile(QOpenGLShader::Vertex, currentVertFile);
+        //m_program->addShaderFromSourceFile(QOpenGLShader::Fragment, currentFragFile);
+        //m_program->link();
+        return false;
+    }
+
+    currentVertFile = vertfilepath;
+    m_posAttr = m_program->attributeLocation("posAttr");
+    Q_ASSERT(m_posAttr != -1);
+    m_colAttr = m_program->attributeLocation("colAttr");
+    Q_ASSERT(m_colAttr != -1);
+    m_matrixUniform = m_program->uniformLocation("matrix");
+    Q_ASSERT(m_matrixUniform != -1);
+
+    m_normAttr = m_program->attributeLocation("normAttr");
+    m_uvAttr = m_program->attributeLocation("uvAttr");
+
+    m_modelviewUniform = m_program->uniformLocation("modelview");
+    m_normalUniform = m_program->uniformLocation("normalMat");
+
+    m_lightPosUniform = m_program->uniformLocation("uLightPos");
+    m_uKa = m_program->uniformLocation("uKa");
+    m_uKd = m_program->uniformLocation("uKd");
+    m_uKs = m_program->uniformLocation("uKs");
+    m_uSpecularColor = m_program->uniformLocation("uSpecularColor");
+    m_uShininess = m_program->uniformLocation("uShininess");
+    /*
+    m_posAttr = newProgram->attributeLocation("posAttr");
```

Comment:
Same as above, can this just be deleted?

Reply:
For some reason the code only works when I reassign the uniform variables every time we load a shader. I will look into this, but for now removing it would probably break the code.

Reply:
I just meant the commented out parts.